app/code/Magento/InventoryCatalog/Model/UpdateLegacyStockStatusByDefaultSourceItem.php

```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Model;
+
+use Magento\Catalog\Model\ProductIdLocatorInterface;
+use Magento\CatalogInventory\Api\Data\StockStatusInterface;
+use Magento\Framework\App\ResourceConnection;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+use Magento\InventoryCatalog\Api\DefaultSourceProviderInterface;
+use Magento\InventoryCatalog\Api\DefaultStockProviderInterface;
+
+/**
+ * Legacy update cataloginventory_stock_status by plain MySql query.
+ * Use for skip save by \Magento\CatalogInventory\Model\ResourceModel\Stock\Item::save
+ */
+class UpdateLegacyStockStatusByDefaultSourceItem
+{
+    /**
+     * @var ResourceConnection
+     */
+    private $resourceConnection;
+
+    /**
+     * @var DefaultSourceProviderInterface
+     */
+    private $defaultSourceProvider;
+
+    /**
+     * @var ProductIdLocatorInterface
+     */
+    private $productIdLocator;
+    /**
+     * @var DefaultStockProviderInterface
+     */
+    private $defaultStockProvider;
+
+    /**
+     * @param ResourceConnection $resourceConnection
+     * @param DefaultSourceProviderInterface $defaultSourceProvider
+     * @param DefaultStockProviderInterface $defaultStockProvider
+     * @param ProductIdLocatorInterface $productIdLocator
+     */
+    public function __construct(
+        ResourceConnection $resourceConnection,
+        DefaultSourceProviderInterface $defaultSourceProvider,
+        DefaultStockProviderInterface $defaultStockProvider,
+        ProductIdLocatorInterface $productIdLocator
+    ) {
+        $this->resourceConnection = $resourceConnection;
+        $this->defaultSourceProvider = $defaultSourceProvider;
+        $this->defaultStockProvider = $defaultStockProvider;
+        $this->productIdLocator = $productIdLocator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function execute(SourceItemInterface $sourceItem)
+    {
+        if ($sourceItem->getSourceId() != $this->defaultSourceProvider->getId()) {
+            return;
+        }
+
+        $productIds = $this->productIdLocator->retrieveProductIdsBySkus([$sourceItem->getSku()]);
```

Comment:
Pls merge develop branch
We have provided global refactoring of product ids by sku resolving

Pls use `\Magento\InventoryCatalog\Model\GetProductIdsBySkusInterface`